SyGuS streaming solution mode

ajreynol opened the pull request:
This pull requests adds support for "streaming solutions mode" for SyGuS (option::sygusStream).  This allows CVC4 to enumerate a stream of solutions to synthesis conjectures instead of just terminating after finding one.  This feature is of interest to invariant generation in Kind2.

It also substantially cleans up the CegConjecture class so that the interface is properly documented and the proper members are encapsulated.  This class is moved to a new file in this commit.

ajreynol commented:
I've also added some optimizations and a bug fix reported by Daniel Larraz for Lustre invariant synthesis problems in my latest commit.  
See CVC4-092517-3-sygusStream at https://www.starexec.org/starexec/secure/details/job.jsp?id=24487

ajreynol commented:
For some reason this is not compiling.

In my last commit, I ran into issues with a "remote rejected" message:

ajreynol@r-lnx214264[~/cvc4-pr-ajr]% git push
Counting objects: 17, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (17/17), done.
Writing objects: 100% (17/17), 3.86 KiB | 0 bytes/s, done.
Total 17 (delta 15), reused 0 (delta 0)
remote: Resolving deltas: 100% (15/15), completed with 15 local objects.
remote: fatal error in commit_refs
To https://github.com/ajreynol/CVC4.git
 ! [remote rejected] sygusStream -> sygusStream (failure)
error: failed to push some refs to 'https://github.com/ajreynol/CVC4.git'


I got around this error by updating my fork into master.  However now it seems that TravisCI fails, but I can't assess why it is failing from the logs.  

I've made a fresh copy of the commits from this branch here to see if things are fixed:  https://github.com/ajreynol/CVC4/tree/sygusStream-v2

ajreynol commented:
Never mind, I had a typo in my previous commit....the push issues were not related.

Independently it is annoying that this does not show up as a hard failure when I do make regress and the TravisCI log is not informative.  Ideally it'd be easier than having to read the 20k line travis log. :)

aniemetz commented:
Yes, this is annoying. Maybe we should make a joint effort to clean up the log to be more informative.
Did you resolve the push issues?

ajreynol commented:
I still don't know what caused the push issues.  But they weren't related to the TravisCI failure.

I'll send an email to cvc4-devel if I run into the same issues again.

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+#include "theory/quantifiers/first_order_model.h"
+#include "theory/quantifiers/term_database_sygus.h"
+#include "theory/theory_engine.h"
+#include "prop/prop_engine.h"
```
Put prop/... after options/... (alphabetize).

aniemetz commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+  delete d_ceg_pbe;
+}
+
+Node CegConjecture::convertToEmbedding( Node n, std::map< Node, Node >& synth_fun_vars, std::map< Node, Node >& visited ){
```
Use unordered_map instead of map for synth_fun_vars, visited.

mpreiner replied:
Is the order of the elements in maps `synth_fun_vars` and `visited` needed? If not why not use an std::unordered_map?

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+namespace theory {
+namespace quantifiers {
+
+void collectDisjuncts( Node n, std::vector< Node >& d ) {
```
This may become a problem if you have a lot of nested OR nodes since there is no cache to check which OR nodes have already been visited.

mpreiner replied:
After discussing this with Andy, the ORs get flattened by the rewriter, so there should not be a problem with deeply nested ORs.

Add a comment to clarify that recursion without a cache is not a problem here.

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+  }
+}
+
+CegConjecture::CegConjecture( QuantifiersEngine * qe, context::Context* c )
```
Is there a reason why context `c` is not used?

ajreynol replied:
At some point I was using c, now I don't need it.  I'd prefer keeping it here since it may come back later

mpreiner replied:
Would it not be better to remove it for now and if you need again later you add it? Otherwise, if you don't need it anymore it will stay there.

barrettcw replied:
I agree - I want to be fairly aggressive about pruning unused code.  It's not hard to add it back if you need it later.

aniemetz commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+  }
+}
+
+void CegConjecture::collectConstants( Node n, std::map< TypeNode, std::vector< Node > >& consts, std::map< Node, bool >& visited ) {
```
Ditto for consts, visited.

mpreiner replied:
Again map vs unordered_map. `visited` could also be an unordered_set.

aniemetz commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+                                         std::vector< Node >& lems ) {
+  Assert( clist.size()==model_values.size() );
+  if( d_ceg_pbe->isPbe() ){
+    //Assert( isGround() );
```
If assertion is not valid, remove.

aniemetz commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+  
+  std::map< Node, Node > csol_active;
+  std::map< Node, std::vector< Node > > csol_ccond_nodes;
+  std::map< Node, std::map< Node, bool > > csol_cpol;    
```
All three maps csol_active, csol_ccond_nodes and csol_cpol seem to be unused?

mpreiner replied:
Are these maps used?

aniemetz commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+  
+  //for conditional evaluation
+  std::map< Node, Node > psubs_visited;
+  std::map< Node, Node > psubs;
```
psubs_visited, psubs: unused?

aniemetz commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+  //add conditional correctness assumptions
+  std::vector< Node > psubs_cond_ant;
+  std::vector< Node > psubs_cond_conc;
+  std::map< Node, std::vector< Node > > psubs_apply;
```
psubs_cond_ant, psubs_cond_conc, psubs_apply: unused?

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+  if( d_stream_guards.empty() ){
+    return Node::null();
+  }else{
+    return d_stream_guards[d_stream_guards.size()-1];
```
Wouldn't `return d_stream_guards.back()` be more readable here?

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.cpp:
```diff
@@ -0,0 +1,797 @@
+            Trace("cegqi-inv-debug") << "...was embedding into grammar." << std::endl;
+          }
+        }else{
+          Trace("cegqi-inv-debug") << sf << " did not used template" << std::endl;
```
Typo: did not _use_ template

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.h:
```diff
@@ -0,0 +1,173 @@
+   * This includes SyGuS inputs where no syntactic restrictions are provided.
+   */
+  bool d_syntax_guided;
+private: // for sygusStream
```
Use only one private block (as discussed yesterday).

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.h:
```diff
@@ -0,0 +1,173 @@
+  std::vector< Node > d_stream_guards;
+  /** get current stream guard */
+  Node getCurrentStreamGuard() const;
+private: //non-syntax guided (deprecated)
```
Same here.

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.h:
```diff
@@ -0,0 +1,173 @@
+private: //non-syntax guided (deprecated)
+  /** the guard for non-syntax-guided synthesis */
+  Node d_nsg_guard;
+private:
```
And here.

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.h:
```diff
@@ -0,0 +1,173 @@
+  bool needsRefinement();
+  /** get the list of candidates */
+  void getCandidateList( std::vector< Node >& clist, bool forceOrig = false );
+public:
```
Same with the public blocks. public blocks should be on top the private section at the end.

mpreiner commented on src/theory/quantifiers/ce_guided_instantiation.cpp:
```diff
@@ -10,18 +10,16 @@
 #include "theory/theory_engine.h"
-#include "prop/prop_engine.h"
+#include "theory/quantifiers/term_database_sygus.h"
+//FIXME
```
What is the FIXME for? Please add a comment on what should be fixed.

mpreiner commented on src/theory/quantifiers/ce_guided_instantiation.h:
```diff
@@ -19,135 +19,13 @@
 
 #include "context/cdchunk_list.h"
```
Is CDChunkList used here? If not remove the `#include`.

mpreiner commented on src/theory/quantifiers/ce_guided_single_inv.h:
```diff
@@ -181,6 +179,15 @@ class CegConjectureSingleInv {
   // the template argument for each function to synthesize (occurs in exactly one position of its template)
   std::map< Node, Node > d_templ_arg;
+  
+ public:
```
Again, merge public blocks and add comments to separate methods. Move public section on top.

ajreynol commented:
I've addressed most of the comments in the review.

I am still holding off on converting map -> unordered_map in old code (the code with std::map here was just moved around but was unmodified) since I am developing a script to do a global replace at some point.  I'd like to isolate and confirm how much the performance gain overall is.

mpreiner commented on src/theory/quantifiers/ce_guided_conjecture.h:
```diff
@@ -0,0 +1,171 @@
+#ifndef __CVC4__THEORY__QUANTIFIERS__CE_GUIDED_CONJECTURE_H
+#define __CVC4__THEORY__QUANTIFIERS__CE_GUIDED_CONJECTURE_H
+
+#include "context/cdchunk_list.h"
```
Seems like this header is also not used.

ajreynol commented:
I've addressed the latest round of comments in my latest commit.
